Index: src/main/java/com/cgvsu/Main.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/com/cgvsu/Main.java b/src/main/java/com/cgvsu/Main.java
--- a/src/main/java/com/cgvsu/Main.java	
+++ b/src/main/java/com/cgvsu/Main.java	
@@ -1,8 +1,6 @@
 package com.cgvsu;
 
 
-import ru.cs.vsu.cg2024.matrix.typesMatrix.Matrix4D;
-
 public class Main {
     public static void main(String[] args) {
         Simple3DViewer.main(args);
